server/src/crm/crm.service.ts

```diff
@@ -32,158 +31,158 @@ export class CrmService {
+    }, {})
+
+    return convertQueryObjectToURL(truthyKeyedObject);
   }
+}
+
+export function coerceToNumber(numericStrings) {
+  return numericStrings
+    // filter out blank strings and undefined, which aren't meaningfully
+    // coercible in CRM
+    .filter(stringish => stringish !== '' && stringish !== undefined)
+    .map(stringish => {
+      // smelly; but let's prefer actual null
+      // coercing 'null' turns to 0, which we don't
+      // want in the API query
+      if (stringish === null) return stringish;
+
+      // Coercing an empty string into a number returns
+      // NaN, which, although a number, is a Double in CRM
+      // which typically expects an Int
+      if (stringish === '') return stringish;
+
+      return Number(stringish);
+    });
+}
 
-  private fixLongODataAnnotations (dataObj) {
-    const newObj = {};
+export function coerceToDateString(epoch) {
+  const date = new Date(epoch * 1000);
 
-    for (let name in dataObj) {
-      const formattedValuePrefix = name.indexOf("@OData.Community.Display.V1.FormattedValue");
-      const logicalNamePrefix = name.indexOf("@Microsoft.Dynamics.CRM.lookuplogicalname");
-      const navigationPropertyPrefix = name.indexOf("@Microsoft.Dynamics.CRM.associatednavigationproperty");
+  return date;
+}
 
-      if (formattedValuePrefix >= 0) {
-        const newName = name.substring(0, formattedValuePrefix);
-        if(newName) newObj[`${newName}_formatted`] = dataObj[name];
-      }
+export function mapInLookup(arrayOfStrings, lookupHash) {
+  return arrayOfStrings.map(string => lookupHash[string]);
+}
 
-      else if (logicalNamePrefix >= 0) {
-        const newName = name.substring(0, logicalNamePrefix);
-        if(newName) newObj[`${newName}_logical`] = dataObj[name];
-      }
+export function all(...statements): string {
+  return '(' + statements
+    .filter(Boolean)
+    .join(' and ') + ')';
+}
 
-      else if (navigationPropertyPrefix >= 0) {
-        const newName = name.substring(0, navigationPropertyPrefix);
-        if (newName) newObj[`${newName}_navigationproperty`] = dataObj[name];
-      }
+export function any(...statements): string {
+  return `(${(statements.join(' or '))})`;
+}
 
-      else {
-        newObj[name] = dataObj[name];
-      }
+// convenience function that takes an iterable and inserts commas
+export function list(...statements): string {
+  return `${(statements.join(','))}`;
+}
+
+export function comparisonOperator(propertyName, operator, value) {
+  let typeSafeValue = value
+
+  if (typeof value === 'string') {
+    if (value !== 'false' && value !== 'true') {
+      typeSafeValue = `'${value}'`;
     }
+  }
+
+  // most likely means it's a date. we want the date formatting that
+  // json stringify provides.
+  if (typeof value === 'object') {
+    const stringyDate = JSON.stringify(value).replace(/"/g, "'");
 
-    return newObj;
+    typeSafeValue = `${stringyDate}`;
   }
 
-  async get(query, maxPageSize = 100, headers= {}) {
```

Comment:
I was hoping to migrate the _utils/crm-web-api.ts file into this crm.service.ts file so that those functions can be part of the nest framework, and we would have less ad-hoc looking util files. Any reason to reintroduce  `_utils/crm-web-api.ts` and move functions like `get()` out of this service?

Reply:
Oh I see,  get() was moved higher up in this file and it's now wrapping the crm-web-api.ts get() .

Reply:
My original reasoning was that since we didn't write any of that code, we should isolate it somewhere and not touch it. But I think I see your point. I'll change it back and I think it'll make for a nicer looking diff.

Reply:
Also wil encourage us to make the code better!

Reply:
@godfreyyeung yep but I did take your advice and move these methos back.

I think if we're to touch these methods more (we will), we _have_ to refactor them. They're very weird. Thoughts?

Reply:
I see your point about intentionally isolating strange code. I agree, if we modify it more we should do some heavy duty refactoring.